rfc/cdevents-spinnaker.md

```diff
@@ -0,0 +1,149 @@
+
+## Overview
+
+Many organisations use more than one CI/CD tools as part of their pipelines and most of the times the organisations implement their own glue code in order to make the tools work with each other.
+
+Implementing [CDEvents](https://cdevents.dev/) to Spinnaker enables Spinnaker to communicate with other CI/CD tools with in the ecosystem. [CDEvents](https://cdevents.dev/) is a common specification for Continuous Delivery events based on CloudEvents.
+Spinnaker can produce and consume [CDEvents](https://cdevents.dev/) using [Java SDK](https://github.com/cdevents/sdk-java) which is developed and maintained by the [CDEvents community](https://cdevents.dev/community/).
+
+## Goals and Non-Goals
+Goals:
+- Event-Driven Standardization with CI/CD tools like Keptn, Tekton, Jenkins, Spinnaker, etc..
+- Spinnaker can able to consume and produce CDEvents from/to `events-broker`
+- One needs to subscribe to `events-broker` with the type of event that needs to consume
+
+Non-Goals:
+- Creation of API end points to produce and consume CDEvents
+
+## Motivation and Rationale
+
+- Currently there is no common specification for CI/CD tools to communicate with in the CI/CD ecosystem. 
+
+- In a complex and fast-moving CI/CD world with a lot of different tools and platforms that need to communicate with each other.
+
+- Multiple pipelines can be run between CI/CD tools by sharing the artifacts and its metadata.
+
+- Increase in traceability with in the ecosystem from commit to deployment of an artifact.
+
+- Operators can deploy Spinnaker with `CDEvents` to run multiple pipelines using different CI/CD tools in ecosystem.
+
+## Timeline
+Proof of concept with Spinnaker, Keptn and Tekton CI/CD tools developed to communicate with each other about occurrences and running pipelines from building an artifact to deploy.
+
+The CDEvents integration process will be achieved in two Goals:
+Goal-1 : Introducing CDEvents to Spinnaker using Spinnaker-Webhooks
+Goal-2 : Integrating CDEvents to Spinnaker using Spinnaker Pub/Sub Automated triggers type(google,amazon) and Notifications
+
+The final implementation aligning with `CDEvents` vocabulary specification will take from 4-6 weeks.
+
+## Design
+
+CDEvents is build upon CloudEvent, the event received will be in CloudEvent format with additional CI/CD specific extensions added to it.
+CDEvent specifications can be found [here](https://cdevents.dev/)
+
+The CDEvents integration process will be achieved in two Goals:
+### Goal-1 : 
+Introducing `CDEvents` to Spinnaker using Spinnaker-Webhook:
+
+Using an existing Spinnaker Webhook `CDEvents` will be consumed/produced and Spinnaker can communicate with other CI/CD tools using CDEvents.
+
+### Consume CDEvents:
+
+A Source `cdevent` will be configured for Webhook type Automated triggers from Spinnaker dashboard.
+The URL `http://<HOST>:8184//webhooks/webhook/cdevent` formed will be placed as subscriber uri to the `events-broker` with specific event type to receive.
```

Comment:
How do we envision a CI and CD system talking? Is it through a single event bus where every system would be talk to it or would each system, CI and CD, have their own bus?

A single event bus for every system is pretty scary tbh. And if we decide on multiple event buses, maybe it makes sense for webhooks to be peer to peer.

Reply:
In the PoC we demonstrated using single knative events-broker used by every system with in the CI/CD ecosystem (https://youtu.be/92Txy-USHg4), 
I think we can have multiple event-brokers but those needs an infrastructure level configuration to pass the events between the different event-brokers used for CI/CD systems.

With this design proposal, webhook can be used for Knative like event-brokers  and send a CloudEvent format using HTTP Post request [cdevent-produce-webhook](https://github.com/spinnaker/governance/blob/0c57e8853b33bc6a85ab113c343397075ed1fed8/rfc/cdevent-produce-webhook.png)